### ISSUE TITLE
Use the pre-trained model

### ISSUE THREAD

**Issue by raialvaro** [gabrieltseng/solar-panel-segmentation#8]
Hi,

Thanks for sharing your code. It's fantastic.

I am new to the world of deep learning.

I have managed to run his coido but I can not prove it with other images.

What should I do?
These are the files created after running all the code:

![image](https://user-images.githubusercontent.com/71493478/93581534-93074700-f9a1-11ea-8bc8-bb5843560df6.png)

Thank so much!

**Comment by gabrieltseng** [comment]
Hi @raialvaro - thanks, I am glad its useful!

It seems like you have successfully run the models - are you asking about how these results can be plotted?

Here is the code snippet I used to take the segmenter predictions and compare them to the input images. I took 3 random images from the prediction dataset and plotted them.

```python
import numpy as np
import matplotlib.pyplot as plt

from solarnet.datasets.utils import denormalize

images = np.load("data/models/segmenter_images.npy")
preds = np.load("data/models/segmenter_preds.npy")
true = np.load("data/models/segmenter_true.npy")

fig, ax = plt.subplots(2, 3, figsize=(15, 10))
ax[0][0].set_ylabel('Ground Truth', size=13)
ax[1][0].set_ylabel('Predictions', size=13)
for ax_idx, im_idx in enumerate(np.random.randint(0, len(preds), 3)):
    ax[0][ax_idx].imshow(np.moveaxis(denormalize(images[im_idx]), 0, -1))
    ax[0][ax_idx].imshow(true[im_idx], alpha=0.5, cmap='Purples')
    ax[1][ax_idx].imshow(np.moveaxis(denormalize(images[im_idx]), 0, -1))
    ax[1][ax_idx].imshow(preds[im_idx], alpha=0.5, cmap='Purples')
plt.savefig('diagrams/segmentation_predictions.png', dpi=300, bbox_inches='tight')

# I ran this in a jupyter notebook, so I also plotted the results there
plt.show()
```

This yields the [diagram in the readme](https://github.com/gabrieltseng/solar-panel-segmentation/blob/master/diagrams/segmentation_predictions.png).

**Comment by raialvaro** [comment]
Hi @gabrieltseng 

This code that you have passed to me also helps, thank you.

But I was referring to uploading any image from a satellite, like sentinel-2, landsat8.

The process would be the following:

1. upload a satellite image to a variable, such as a sentinel-2, landsat 8 image that has solar panels

2. resize, normalize the image to what the model needs to predict

2. load the trained model.

3. insert the resized image to the model.

4. check if it performs the segmentation perfectly as with the code that you passed previously


I hope I have explained well :)

Thank so much!

**Comment by gabrieltseng** [comment]
Hi @raialvaro ,

Unfortunately, I don't think this model will immediately transfer to sentinel or landsat imagery - the training imagery used here was collected from USGS imagery with a minimum resolution of 30cm (https://www.nature.com/articles/sdata2016106).

Sentinel-2 data has a 10m resolution and Landsat-8 data has a resolution of 30m, which is orders of magnitude lower. Your best bet would be to retrain a model specifically on that data, but I am also not sure a solar panel will be easy (/ possible) to detect at that resolution.

**Comment by raialvaro** [comment]
Hi @gabrieltseng 

Thanks for replying again

of course, you are right, Landsat or Sentinel images is a bad idea.

what I wanted to know is: how can I pass an image and segment the solar panels it contains? Or check an image that does not contain solar panels, and effectively does not perform segmentation

I know collected from USGS imagery, so how can I pass an image of the model containing solar panels?
For example, an image from the USGS showing the solar panels of a neighbor's house

I hope you can help :)

Thank you very much

**Comment by gabrieltseng** [comment]
Ah, I understand. You would need to save the image as an npy file (how you do this depends on which format you get the USGS data in, but most geospatial datatypes have a "to numpy" function - [here is one for xarray](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.values.html)). 

One note is that the model expects images to be 224 by 224 pixels, so you would need to ensure this is the case.

Once this is done, you should be able to pass the model to the trained segmenter to get your output. You would just need to normalize it, using the `utils.normalize` function. The code might look something like this (I haven't run this, so it will need to be tweaked):

```python
import torch
import numpy as np

from solarnet.utils import normalize
from solarnet.models import Segmenter


data_folder = "data"

my_model = Segmenter()
my_model.load_state_dict(f"{data_folder}/segmenter.model")

input_image = normalize(np.load(path_to_my_image))
input_tensor = torch.from_numpy(input_image).float()

preds = my_model(input_tensor)
```

Hope this helps!

**Comment by raialvaro** [comment]
Hi @gabrieltseng

Thanks for your help. It has been very useful

I am using PNOA images with 30 cm resolution and 224x224 size as you said

But I have an error when normalizing the numpy file that I can't solve.

I show you the code and the error at the end


```
# load image and show

import numpy as np
import matplotlib.pyplot as plt

from matplotlib import image
from matplotlib import pyplot

image = image.imread('24178_rgb.tif')
print(image.dtype)
print(image.shape)
```
![image](https://user-images.githubusercontent.com/71493478/93978654-92ddc180-fd7c-11ea-8c60-0f761f430f79.png)

```
#transform image file to numpy class

from PIL import Image
from numpy import asarray

image = Image.open('24178_rgb.tif')
data = asarray(image)
print(type(data))
print(data.shape)
```

![image](https://user-images.githubusercontent.com/71493478/93980233-a8ec8180-fd7e-11ea-8c7d-51f0b253040c.png)

```
print(data)
```
![image](https://user-images.githubusercontent.com/71493478/93980331-c6b9e680-fd7e-11ea-855c-73e3e7ae95de.png)

#load model with your code
```
import torch
from solarnet.datasets.utils import normalize
from solarnet.models import Segmenter

data_folder = ".../models"
my_model = Segmenter()
my_model.load_state_dict(torch.load(f"{data_folder}/segmenter.model"))
print(my_model)
```
![image](https://user-images.githubusercontent.com/71493478/93979878-2c59a300-fd7e-11ea-86c5-4fe63298be75.png)

```
input_image = normalize(data)
```
![image](https://user-images.githubusercontent.com/71493478/93979980-51e6ac80-fd7e-11ea-8088-70fd8bce6ca0.png)


I don't understand the problem when I am using a 224x224 image with 3 channels, RGB.

Can you help me?

 Thank you very much

**Comment by gabrieltseng** [comment]
The `normalize` function expects the input image to have shape `[channels, height, width]`, and it looks like you are passing [height, width, channels]`. Moving the last dimension to the first should do the trick.

You also might need to [expand the first dimension](https://numpy.org/doc/stable/reference/generated/numpy.expand_dims.html) when you pass the image to the model, something like

```python
image_as_batch = np.expand_dims(image, axis=0)
```
so that the model sees a batch of size 1 being passed.

Hope this helps!

**Comment by raialvaro** [comment]
Thank so much! @gabrieltseng 

It seems to have worked just as you anticipated

![image](https://user-images.githubusercontent.com/71493478/93986552-c7567b00-fd86-11ea-8562-d10dc2d3beda.png)

I am very happy :)


This "preds" variable is a tensor. If I am not mistaken, I think it really is the image with segmentation?

If so, the next step is to convert "preds" to image format?

I'm sorry for all the inconvenience I'm causing you

Thank!

**Comment by gabrieltseng** [comment]
Yes - one way you could do this is by saving to npy (`tensor.numpy()`) and then using the plotting functions I shared above.

One last note - `reshape` might not be the function you want to be using. [`np.moveaxis`](https://numpy.org/doc/stable/reference/generated/numpy.moveaxis.html) probably makes more sense.

**Comment by raialvaro** [comment]
Hi @gabrieltseng 

Everything went well as I explained previously.
But when I go to show the prediction in matploit it doesn't run and shows this error:

![image](https://user-images.githubusercontent.com/71493478/94039934-9d25ad00-fdc8-11ea-8086-93a0bca80362.png)

I have been observing and the size of the numpy files are different for each one.

![image](https://user-images.githubusercontent.com/71493478/94040257-01e10780-fdc9-11ea-88bc-b38f73300c86.png)

My variable "my_predict_image" variable has a different format than your "images" variable.

I do not know what is happening.

I think I'm going to get it

I would appreciate your help :)

Thank you very much

**Comment by raialvaro** [comment]
Hi @gabrieltseng 

I forgot to tell you to also change the syntax of `np.moveaxis` just like you said.

![image](https://user-images.githubusercontent.com/71493478/94110091-58d6f300-fe42-11ea-8c89-be542fc11c67.png)

I mean the problem doesn't come from that syntax

Thanks again!

**Comment by gabrieltseng** [comment]
Yup, so the image has shape `1, 1, 224, 224` which is `batch, channel, height, width`

The [plt.imshow](https://matplotlib.org/3.2.1/api/_as_gen/matplotlib.pyplot.imshow.html) expects an input of shape `224, 224` - just height and width, since there is only one channel (the prediction). You can get this by doing
```python
my_predict_image = my_predict_image[0, 0, :, :]
```

**Comment by raialvaro** [comment]
Thank you @gabrieltseng 

There were good results using the following syntax for me `np.squeeze(my_predict_image, axis=0)[0]`

![image](https://user-images.githubusercontent.com/71493478/94144734-cea68300-fe71-11ea-9ab9-dda3193f8bfa.png)

But it also works with your code `my_predict_image = my_predict_image[0, 0, :, :]`

![image](https://user-images.githubusercontent.com/71493478/94144772-dbc37200-fe71-11ea-9d0d-3aaac0310ee6.png)

![image](https://user-images.githubusercontent.com/71493478/94145056-41aff980-fe72-11ea-9037-fbd1a2de09c1.png)

I don't interrupt him more times

Thanks for all the help you have really offered me

**Comment by sara713** [comment]
hi
would u please provide the pretrained model for testing?
the data is too huge. all people cant work with it.
